### DISCUSSION TITLE
Could EqualException display the types of object array elements?

### DISCUSSION THREAD

**Discussion by Quahu** [xunit/xunit#2452] [Ideas]
This is prompted by my mistake in #2450.

When comparing individual `object`s the types are provided alongside the values.
Example:
```cs
Assert.Equal<object>(42, 42l); // int, long
```

Yields:
```
Xunit.Sdk.EqualException
Assert.Equal() Failure
Expected: 42 (System.Int32)
Actual:   42 (System.Int64)
```

However with arrays that isn't the case.
Example:
```cs
var array1 = new object[] { 42 }; // int
var array2 = new object[] { 42l }; // long
Assert.Equal(array1, array2);
```
Yields:
```
Xunit.Sdk.EqualException
Assert.Equal() Failure
Expected: Object[] [42]
Actual:   Object[] [42]
```
Which makes it not possible to tell what the types of the values were.
Could the exception display the types of the values in the object array?

Edit:
I guess this applies to `AssertActualExpectedException` and all derived exceptions that can accept collections of `object`s.

**Comment by bradwilson** [comment]
The new assertion message is more helpful:

```
Assert.Equal() Failure
                 ↓ (pos 2)
Expected: [1, 2, 3]
Actual:   [1, 2, 3]
                 ↑ (pos 2)
```

But it still doesn't show type. I'm open to showing type as well.

**Reply by bradwilson** [reply]
https://github.com/xunit/xunit/issues/2458
